### PR TITLE
refactoring_us_752

### PR DESCRIPTION
**Pull request by FranckLecuyer**

### CODE REVIEW

File: src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java [hunk 1]
```diff
@@ -448,4 +474,19 @@ public void doUpdateBuildStatus(UUID nodeUuid, BuildStatus buildStatus) {
...
         return nodesRepository.findById(nodeUuid).map(n -> repositories.get(n.getType()).getBuildStatus(nodeUuid)).orElse(BuildStatus.NOT_BUILT);
     }
+
+    @Transactional(readOnly = true)
+    public Optional<UUID> doGetParentNode(UUID nodeUuid, NodeType nodeType) {
+        NodeEntity nodeEntity = nodesRepository.findById(nodeUuid).orElseThrow(() -> new StudyException(ELEMENT_NOT_FOUND));
+        if (nodeEntity.getType() == NodeType.ROOT || nodeEntity.getParentNode().getType() == nodeType) {
+            return Optional.of(nodeEntity.getParentNode().getIdNode());
+        } else {
+            return doGetParentNode(nodeEntity.getParentNode().getIdNode(), nodeType);
```
Comment (JacquesBorsenberger) [inline src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java:484]: that means that we make one request per node until we get the type requested...

File: src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java [hunk 2]
```diff
@@ -288,7 +297,17 @@ public UUID getStudyRootNodeUuid(UUID studyId) {
 
     @Transactional
     public Optional<UUID> doGetModificationGroupUuid(UUID nodeUuid, boolean generateId) {
-        return nodesRepository.findById(nodeUuid).flatMap(n -> repositories.get(n.getType()).getModificationGroupUuid(nodeUuid, generateId));
+        Optional<NodeEntity> nodeEntity = nodesRepository.findById(nodeUuid);
```
Comment (JacquesBorsenberger) [inline src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java:300]: exactly the same as doGetVariantId, except the last line, can use a (private) function

**Review by JacquesBorsenberger — CHANGES_REQUESTED** [review]
Just the duplication code to move to a function ant this will be ok.

File: src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java [hunk 3]
```diff
@@ -278,7 +277,17 @@ public UUID getStudyRootNodeUuid(UUID studyId) {
...
+        if (!nodeEntity.isPresent()) {
+            throw new StudyException(ELEMENT_NOT_FOUND);
+        }
+        // if nodeUuid is a model node, get parent node of type network modification node
+        UUID tmpUuid = nodeUuid;
+        if (nodeEntity.get().getType() == NodeType.MODEL) {
+            tmpUuid = self.doGetParentNode(nodeUuid, NodeType.NETWORK_MODIFICATION).orElseThrow(() -> new StudyException(ELEMENT_NOT_FOUND));
+        }
+        UUID modificationNodeUuid = tmpUuid;
+        return nodesRepository.findById(modificationNodeUuid).flatMap(n -> repositories.get(n.getType()).getVariantId(modificationNodeUuid, generateId));
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java:290]: No, variantId is only for modification node

File: src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java (hunk 1 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java:484]: Maybe, an optimization could be done later ...

File: src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java (hunk 2 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java]: Code factorized

Comment (JacquesBorsenberger) [reply src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java]: ok

**Review by JacquesBorsenberger — APPROVED** [review]

**Comment by SlimaneAmar** [comment]
I think we have a design problem with AbstractNodeRepositoryProxy
It's an abstract class which implements specific things for each type of node

File: src/main/java/org/gridsuite/study/server/networkmodificationtree/AbstractNodeRepositoryProxy.java [hunk 4]
```diff
@@ -40,27 +40,44 @@ protected AbstractNodeRepositoryProxy(NodeInfoEntityRepository nodeInfoRepositor
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/study/server/networkmodificationtree/AbstractNodeRepositoryProxy.java:40]: See comment about the design

**Comment by SlimaneAmar** [comment]
I thonk we have many methods which return an optional but which do not correspond to an expected or normal state

**Comment by SlimaneAmar** [comment]
Nedd to add a test class only for the build ?

**Review by SlimaneAmar — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/study/server/RestResponseEntityExceptionHandler.java [hunk 5]
```diff
@@ -50,6 +50,8 @@
                     return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(studyException.getMessage());
                 case DELETE_EQUIPMENT_FAILED:
                     return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(studyException.getMessage());
+                case BAD_NODE_TYPE:
+                    return ResponseEntity.status(HttpStatus.FORBIDDEN).body(studyException.getMessage());
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/study/server/RestResponseEntityExceptionHandler.java:54]: Done

File: src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java [hunk 6]
```diff
@@ -388,32 +402,39 @@ private void getSecurityAnalysisResultUuids(UUID nodeUuid, List<UUID> uuids) {
 
     private void getBuildInfos(NodeEntity nodeEntity, BuildInfos buildInfos) {
         AbstractNode node = repositories.get(nodeEntity.getType()).getNode(nodeEntity.getIdNode());
-        if (node.getType() == NodeType.ROOT) {
-            RootNode rootNode = (RootNode) node;
-            if (rootNode.getBuildStatus() != BuildStatus.BUILT && rootNode.getNetworkModification() != null) {
-                buildInfos.insert(rootNode.getNetworkModification());
+        if (node.getType() == NodeType.MODEL) {
+            ModelNode modelNode = (ModelNode) node;
+            if (modelNode.getBuildStatus() != BuildStatus.BUILT) {
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java:407]: No, this method traverses the tree upwards to collect all modifications and the original variant id, so there is no unauthorized type (BAD_NODE_TYPE) to test here ...

File: src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java [hunk 7]
```diff
@@ -278,7 +277,17 @@ public UUID getStudyRootNodeUuid(UUID studyId) {
 
     @Transactional
     public Optional<String> doGetVariantId(UUID nodeUuid, boolean generateId) {
-        return nodesRepository.findById(nodeUuid).flatMap(n -> repositories.get(n.getType()).getVariantId(nodeUuid, generateId));
+        Optional<NodeEntity> nodeEntity = nodesRepository.findById(nodeUuid);
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java:279]: Code has been changed after firt reviewer remarks

File: src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java [hunk 8]
```diff
@@ -278,7 +277,17 @@ public UUID getStudyRootNodeUuid(UUID studyId) {
...
+        if (!nodeEntity.isPresent()) {
+            throw new StudyException(ELEMENT_NOT_FOUND);
+        }
+        // if nodeUuid is a model node, get parent node of type network modification node
+        UUID tmpUuid = nodeUuid;
+        if (nodeEntity.get().getType() == NodeType.MODEL) {
+            tmpUuid = self.doGetParentNode(nodeUuid, NodeType.NETWORK_MODIFICATION).orElseThrow(() -> new StudyException(ELEMENT_NOT_FOUND));
+        }
+        UUID modificationNodeUuid = tmpUuid;
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java]: Done

File: src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java [hunk 9]
```diff
@@ -448,4 +474,19 @@ public void doUpdateBuildStatus(UUID nodeUuid, BuildStatus buildStatus) {
     public BuildStatus getBuildStatus(UUID nodeUuid) {
         return nodesRepository.findById(nodeUuid).map(n -> repositories.get(n.getType()).getBuildStatus(nodeUuid)).orElse(BuildStatus.NOT_BUILT);
     }
+
+    @Transactional(readOnly = true)
+    public Optional<UUID> doGetParentNode(UUID nodeUuid, NodeType nodeType) {
+        NodeEntity nodeEntity = nodesRepository.findById(nodeUuid).orElseThrow(() -> new StudyException(ELEMENT_NOT_FOUND));
+        if (nodeEntity.getType() == NodeType.ROOT || nodeEntity.getParentNode().getType() == nodeType) {
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java:476]: Controls added

File: src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java [hunk 10]
```diff
@@ -276,9 +275,23 @@ public UUID getStudyRootNodeUuid(UUID studyId) {
...
+        if (!nodeEntity.isPresent()) {
+            throw new StudyException(ELEMENT_NOT_FOUND);
+        }
+        // if nodeUuid is a model node, get parent node of type network modification node
+        UUID tmpUuid = nodeUuid;
+        if (nodeEntity.get().getType() == NodeType.MODEL) {
+            tmpUuid = self.doGetParentNode(nodeUuid, NodeType.NETWORK_MODIFICATION).orElseThrow(() -> new StudyException(ELEMENT_NOT_FOUND));
+        }
+        return tmpUuid;
+    }
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java:289]: Done

File: src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java [hunk 11]
```diff
@@ -388,32 +402,39 @@ private void getSecurityAnalysisResultUuids(UUID nodeUuid, List<UUID> uuids) {
 
     private void getBuildInfos(NodeEntity nodeEntity, BuildInfos buildInfos) {
         AbstractNode node = repositories.get(nodeEntity.getType()).getNode(nodeEntity.getIdNode());
-        if (node.getType() == NodeType.ROOT) {
-            RootNode rootNode = (RootNode) node;
-            if (rootNode.getBuildStatus() != BuildStatus.BUILT && rootNode.getNetworkModification() != null) {
-                buildInfos.insert(rootNode.getNetworkModification());
+        if (node.getType() == NodeType.MODEL) {
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java:405]: This is a recursive method

File: src/main/java/org/gridsuite/study/server/networkmodificationtree/AbstractNodeRepositoryProxy.java [hunk 12]
```diff
@@ -40,27 +40,44 @@ protected AbstractNodeRepositoryProxy(NodeInfoEntityRepository nodeInfoRepositor
 
     public abstract NodeDto toDto(NodeInfoEntity node);
 
-    public abstract Optional<String> getVariantId(AbstractNode node, boolean generateId);
+    public Optional<String> getVariantId(AbstractNode node, boolean generateId) {
+        return Optional.empty();
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/study/server/networkmodificationtree/AbstractNodeRepositoryProxy.java:44]: Optional removed

File: src/main/java/org/gridsuite/study/server/networkmodificationtree/AbstractNodeRepositoryProxy.java [hunk 13]
```diff
@@ -40,27 +40,44 @@ protected AbstractNodeRepositoryProxy(NodeInfoEntityRepository nodeInfoRepositor
...
+    }
 
-    public abstract Optional<UUID> getModificationGroupUuid(AbstractNode node, boolean generateId);
+    public Optional<UUID> getModificationGroupUuid(AbstractNode node, boolean generateId) {
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/study/server/networkmodificationtree/AbstractNodeRepositoryProxy.java:47]: Optional removed

File: src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java [hunk 14]
```diff
@@ -276,9 +275,23 @@ public UUID getStudyRootNodeUuid(UUID studyId) {
...
+            tmpUuid = self.doGetParentNode(nodeUuid, NodeType.NETWORK_MODIFICATION).orElseThrow(() -> new StudyException(ELEMENT_NOT_FOUND));
+        }
+        return tmpUuid;
+    }
+
     @Transactional
     public Optional<String> doGetVariantId(UUID nodeUuid, boolean generateId) {
-        return nodesRepository.findById(nodeUuid).flatMap(n -> repositories.get(n.getType()).getVariantId(nodeUuid, generateId));
+        UUID modificationNodeUuid = getModificationNodeUuidFromNode(nodeUuid);
+        return nodesRepository.findById(modificationNodeUuid).flatMap(n -> repositories.get(n.getType()).getVariantId(modificationNodeUuid, generateId));
```
[the remark this answers is not in the document]
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java:294]: Done

**Comment by FranckLecuyer** [comment]
> I think we have a design problem with AbstractNodeRepositoryProxy It's an abstract class which implements specific things for each type of node

In another US

File: src/main/java/org/gridsuite/study/server/networkmodificationtree/AbstractNodeRepositoryProxy.java (hunk 4 above)
Comment (FranckLecuyer) [reply src/main/java/org/gridsuite/study/server/networkmodificationtree/AbstractNodeRepositoryProxy.java:40]: In another US

Comment (SlimaneAmar) [reply src/main/java/org/gridsuite/study/server/networkmodificationtree/AbstractNodeRepositoryProxy.java:39]: Yes

File: src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (SlimaneAmar) [reply src/main/java/org/gridsuite/study/server/NetworkModificationTreeService.java:279]: not done ?

**Review by SlimaneAmar — APPROVED** [review]

**Pull request merged**